src/hooks/usePing.ts

```diff
@@ -0,0 +1,69 @@
+import { useEffect, useState } from 'react';
+import { ping } from '../utils';
+interface IntranetCheck {
+    isSourceAvailable: boolean;
+}
+
+interface CachedPing {
+    pingTask: Promise<boolean>;
+    timestamp: Date;
+}
+
+const cachedPings: Record<string, CachedPing> = {};
+
+function dateDifferenceInSeconds(date1?: Date, date2?: Date): number {
+    if (!date1 || !date2) {
+        return 9999999;
+    }
+    date2 = new Date(date2); //typescript doesn't know the difference between string and date...
+    date1 = new Date(date1);
+    const diff = Math.abs(date1.getTime() - date2.getTime());
+    return Math.ceil(diff / 1000);
+}
+
+async function pingSourceWithCache(
```

Comment:
Should be extracted to it's own logic file, and tested seprately.

Reply:
I think that extracting this private function isn't justified: it's only used by the usePing hook in this file, and with that it's quite specific.

I would keep it here, hidden away as it's not used by others just yet - once that changes, then I think it would be more justified to move this one out to it's own file.